### PR TITLE
Demo

### PR DESCRIPTION
**Pull request by mher96**

### CODE REVIEW

File: src/tag.api.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
+import {AxiosInstance} from 'axios'
+import {TagsResponse} from './interfaces/mainInterfaces'
+
+export const createTags = async (
+    spireClient: AxiosInstance, 
+    journalId: string,
+    entryId: string,
+    tag: string
+): Promise<string[]> =>{
```
Comment (artakhnoyan) [inline src/tag.api.ts]: Bad practice to return strings array, cause if you consider to add some property to Tag (for example created_date), you need to change api.

File: src/interfaces/mainInterfaces.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
+import {AxiosInstance} from 'axios'
```
Comment (artakhnoyan) [inline src/interfaces/mainInterfaces.ts]: bad naming.

File: src/index.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+import getUser from './user.api'
+import {
+    testPing,
+    createBugoutUser,
+    verifyBugoutUser,
+    setBugoutUserToken,
+    loginBugoutUser,
+} from './bugout.api'
+
+const BugoutClient = (
```
Comment (artakhnoyan) [inline src/index.ts]: BugoutClient should have its own interface that will define API

File: src/interfaces/mainInterfaces.ts (hunk 2 above)
Comment (artakhnoyan) [reply src/interfaces/mainInterfaces.ts]: Also these are types, not interfaces.

File: src/journal.api.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
+import {AxiosInstance} from 'axios'
```
Comment (artakhnoyan) [inline src/journal.api.ts]: The best way to provide API is interfaces. I suggest to create an interface related to journal (or anything else) and export Interface reference.

File: src/tag.api.ts (hunk 1 above)
Comment (artakhnoyan) [reply src/tag.api.ts]: Also It will be more readable.

File: src/journal.api.ts (hunk 4 above)
Comment (artakhnoyan) [reply src/journal.api.ts]: Also we can create `index.ts` and `journal.ts` in `journal` folder and export `Journal` interface implementation in `index.ts`

File: src/bugout.api.ts [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+const hanldeError = (e: any) => {
+    if(e.response) {
+        return {
+            ...e.response,
+            status: e.response.status,
+        }
+    }
+}
+
+export const testPing = async (broodClient: AxiosInstance, spireClient: AxiosInstance) => {
```
Comment (artakhnoyan) [inline src/bugout.api.ts]: Do we have the opportunity to restrict import of the files that we use for local implementation?

File: src/index.ts [hunk 6]
```diff
@@ -0,0 +1,14 @@
+import {BROODURL, SPIREURL} from './constants'
+import BugoutClient from './bugoutClient'
+import {ApiUrls} from './bugoutClient/bugoutClient'
+
+
+const module = (
+    ApiUrls: ApiUrls = {broodURL: BROODURL, spireURL: SPIREURL},
+    bugoutToken?: string,
+    bugoutClientID?: string
+) => (
```
Comment (gogjhan) [inline src/index.ts]: missing interface

File: src/index.ts [hunk 7]
```diff
@@ -0,0 +1,14 @@
+import {BROODURL, SPIREURL} from './constants'
+import BugoutClient from './bugoutClient'
+import {ApiUrls} from './bugoutClient/bugoutClient'
+
+
+const module = (
+    ApiUrls: ApiUrls = {broodURL: BROODURL, spireURL: SPIREURL},
```
Comment (gogjhan) [inline src/index.ts]: receive all options as object argument

Comment (gogjhan) [reply src/index.ts]: change to `apiUrls, token, clientID`

File: src/index.ts [hunk 8]
```diff
@@ -0,0 +1,14 @@
+import {BROODURL, SPIREURL} from './constants'
+import BugoutClient from './bugoutClient'
```
Comment (gogjhan) [inline src/index.ts]: rename bugoutClient folder to just client, have `index.js types.js` files in it

File: src/bugoutClient/bugoutClient.ts [hunk 9]
```diff
@@ -0,0 +1,19 @@
+export interface User {
```
Comment (gogjhan) [inline src/bugoutClient/bugoutClient.ts]: these are types not interfaces

File: src/bugoutClient/index.ts [hunk 10]
```diff
@@ -0,0 +1,123 @@
+import axios, {AxiosInstance} from 'axios'
+import * as querystring from 'querystring'
+import {User, Auth, ApiUrls} from './bugoutClient'
+import getUser from '../user'
+
+
+const hanldeError = (e: any) => {
```
Comment (gogjhan) [inline src/bugoutClient/index.ts]: create `helper.js` relative to this file, and include it inside the helper.js

Comment (gogjhan) [reply src/bugoutClient/index.ts]: typo `handleError`

File: src/bugoutClient/index.ts [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+            status: e.response.status,
+        }
+    }
+}
+
+const BugoutClient = (
+    ApiUrls: ApiUrls,
+    bugoutToken?: string,
+    bugoutClientID?: string, 
+    ) => {
```
Comment (gogjhan) [inline src/bugoutClient/index.ts]: missing interface

File: src/bugoutClient/index.ts [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+            code: results[0].status,
+        },
+        spire: {
+            url: results[1].config.baseURL,
+            code: results[1].status,
+        }
+    }
+    }
+    const createUser = async (
+        email: User['email'],
```
Comment (gogjhan) [inline src/bugoutClient/index.ts]: change to just `string`, change everywhere

File: src/bugoutClient/index.ts (hunk 11 above)
Comment (gogjhan) [reply src/bugoutClient/index.ts]: example

Screen Shot 2020-12-03 at 16 41 04

File: src/bugoutClient/index.ts [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+            spireClient.defaults.headers['authorization'] = `Bearer ${token}`
+            
+            const personalInfoResponse = await broodClient.get<User>('/user')
+            personalInformation = personalInfoResponse.data
+            authInfo = {
+                token_type: 'bearer',
+                access_token: token,
+            }
+            
+            return true
```
Comment (gogjhan) [inline src/bugoutClient/index.ts]: do not return anything, should be `void`

File: src/bugoutClient/index.ts [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+        try {
+            const userResponse = await broodClient.post<User>('/user', registrationForm)
+
+            return userResponse.data;
+        } catch(e) {
+            throw e.response.data
+        }
+    }
+    const verify = async (code: string): Promise<User> => {
+        if(!personalInformation || !authInfo) throw new Error('Token required, configurate token using setToken function')
```
Comment (gogjhan) [inline src/bugoutClient/index.ts]: remove this if, function will handle itself

File: src/bugoutClient/index.ts [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+
+const BugoutClient = (
+    ApiUrls: ApiUrls,
+    bugoutToken?: string,
+    bugoutClientID?: string, 
+    ) => {
+// ToDo here must be request to auth api
+    const broodClient: AxiosInstance = axios.create({baseURL: ApiUrls.broodURL})
+    const spireClient: AxiosInstance = axios.create({baseURL: ApiUrls.spireURL})
+    let personalInformation: User
```
Comment (gogjhan) [inline src/bugoutClient/index.ts]: remove

File: src/bugoutClient/index.ts [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+    let authInfo: Auth
```
Comment (gogjhan) [inline src/bugoutClient/index.ts]: remove

File: src/user/index.ts [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+    updateEntry,
+    deleteEntry,
+    getEntriesByJournal,
+    searchEntriesByJournal,
+} from '../entry'
+import {Journal} from '../journal/journal'
+import {Entry, EntryMutable} from '../entry/entry'
+import {Tag, TagDescription} from '../tag/tag'
+
+const getUser = (spireClient: AxiosInstance) => (
```
Comment (gogjhan) [inline src/user/index.ts]: missing interface

File: src/user/index.ts [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+    {
```
Comment (gogjhan) [inline src/user/index.ts]: let's have something like this
```js
const getUser = (spireClient: AxiosInstance): Interface => (
    {
        ...createJournalsAPI(spireClient),
        ...createEntriesAPI(spireClient),
        ...createTagsAPI(spireClient),
    }
);
```

File: src/bugoutClient/index.ts [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+        }
+    }
+
+    return {
+            ping,
+            createUser,
+            verify,
+            login,
+            setToken,
+            user: () => getUser(spireClient),
```
Comment (gogjhan) [inline src/bugoutClient/index.ts]: rename function to `createUserAPI`

File: src/tag/index.ts [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+import {Tag, TagDescription} from './tag'
+import {Entry} from '../entry/entry'
+import {Journal} from '../journal/journal'
+
+export const createTags = async (
+    spireClient: AxiosInstance, 
+    journalId: Journal['id'],
+    entryId: Entry['id'],
+    tag: Tag['name']
+): Promise<string[]> =>{
```
Comment (gogjhan) [inline src/tag/index.ts]: change to `Promise<Response<Tag[]>>` change everywhere

- [ ] do not return array of primitive types, always objects
- [ ] all function should return same response object

File: src/tag/index.ts [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+    try { 
+        const entryResponse = await spireClient.post<string[]>(`/journals/${journalId}/entries/${entryId}/tags`, {tags: tag})
+        return entryResponse.data
```
Comment (gogjhan) [inline src/tag/index.ts]: change to
```js
return {
            data: entryResponse.data.map(
                (tag) => ({
                    name: tag,
                })
            )
        }
```

File: src/interfaces/mainInterfaces.ts [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+
+export type ApiUrls = {
```
Comment (gogjhan) [inline src/interfaces/mainInterfaces.ts]: remove this file

**Pull request merged**